### PR TITLE
More preparations for #361

### PR DESCRIPTION
**Pull request by Nadrieril**

### CODE REVIEW

File: charon/src/ast/types/vars.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    Ord,
+    Serialize,
+    Deserialize,
+    Drive,
+    DriveMut,
+)]
+#[serde(transparent)]
+pub struct DeBruijnId {
+    pub index: usize,
+}
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply charon/src/ast/types/vars.rs:26]: In my mind the `generate_index_type!` types are meant for indexing into a `Vector` we never manipulate the underlying integer, they may as well be opaque. Here I keep manipulating the integer value as an integer, so it feels different. Note in particular that there's no `impl Idx for DeBruijnId`, which is the main purpose of `generate_index_type!`.

[the remark this answers is not in the document]
Comment (sonmarcho) [reply charon/src/ast/types/vars.rs:26]: Yes, that makes sense.

File: charon/src/ast/types/vars.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    pub span: Option<Span>,
+    /// Where the predicate was written, relative to the item that requires it.
+    #[charon::opaque]
+    pub origin: PredicateOrigin,
+    /// The trait that is implemented.
+    #[charon::rename("trait")]
+    pub trait_: PolyTraitDeclRef,
+}
+
+pub type RegionDbVar = DeBruijnVar<BoundRegionId, FreeRegionId>;
```
Comment (sonmarcho) [inline charon/src/ast/types/vars.rs:116]: I just realized I'm not super happy with the name `DeBruijnVar`: to me a de Bruijn var is necessarily a bound var, while here it actually refers to a bound var or a free var. I'm aware the name `Var` is already taken. Maybe we should use `VarDecl` (for the current `Var` type) and `Var` (for the current `DeBruijnVar`)?

File: charon/src/ast/types/vars.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+///                                   Var id: 1
+/// ```
+#[derive(Debug, PartialEq, Eq, Copy, Clone, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub enum DeBruijnVar<Bound, Free> {
+    /// A variable attached to the nth binder, counting from the inside.
+    Bound(DeBruijnId, Bound),
+    /// A variable attached to an implicit binder outside all other binders. This is not present in
+    /// translated code, and only provided as a convenience for convenient variable manipulation.
+    Free(Free),
+}
```
Comment (sonmarcho) [inline charon/src/ast/types/vars.rs:57]: I just realized I'm not super happy with the name `DeBruijnVar`: to me a de Bruijn var is necessarily a bound var, while here it actually refers to a bound var or a free var. I'm aware the name `Var` is already taken. Maybe we should use `VarDecl` (for the current `Var` type) and `Var` (for the current `DeBruijnVar`)?

Comment (sonmarcho) [reply charon/src/ast/types/vars.rs:57]: (and yes, it does make sense to see a free var as bound outside all binders: it's just not the standard interpretation I believe)

Comment (Nadrieril) [reply charon/src/ast/types/vars.rs:57]: I was thiking the same. The current `Var` should be called `Local` anyway, I thought I had renamed it already.

Comment (Nadrieril) [reply charon/src/ast/types/vars.rs:57]: I'll add it to the [list of renames](https://github.com/AeneasVerif/charon/issues/287)

**Pull request merged**